hack/generator/pkg/codegen/storage/storage_type_factory.go

```diff
@@ -32,67 +35,77 @@ func NewStorageTypeFactory(group string, idFactory astmodel.IdentifierFactory) *
 
 	result := &StorageTypeFactory{
 		group:             group,
-		inputTypes:        types,
+		types:             types,
+		specTypes:         astmodel.NewTypeNameSet(),
+		statusTypes:       astmodel.NewTypeNameSet(),
 		idFactory:         idFactory,
 		conversionMap:     make(map[astmodel.PackageReference]astmodel.PackageReference),
 		functionInjector:  NewFunctionInjector(),
 		resourceHubMarker: NewHubVersionMarker(),
-		typeConverter:     NewTypeConverter(types),
+		typeConverter:     NewTypeConverter(types, idFactory),
 	}
 
 	return result
 }
 
 // Add the supplied type definition to this factory
-func (f *StorageTypeFactory) Add(def astmodel.TypeDefinition) {
-	f.inputTypes.Add(def)
+func (f *StorageTypeFactory) Add(definitions ...astmodel.TypeDefinition) {
+	for _, def := range definitions {
+		f.types.Add(def)
+
+		if rt, ok := astmodel.AsResourceType(def.Type()); ok {
+			// We have a resource type
+			if tn, ok := astmodel.AsTypeName(rt.SpecType()); ok {
+				// Keep track of all our spec types
+				f.specTypes.Add(tn)
+			}
+			if tn, ok := astmodel.AsTypeName(rt.StatusType()); ok {
+				// Keep track of all our status types
+				f.statusTypes.Add(tn)
+			}
+		}
+	}
 }
 
-// Types returns inputTypes contained by the factory, including all new storage variants and modified
-// api inputTypes. If any errors occur during processing, they're returned here.
+// Types returns types contained by the factory, including all new storage variants and modified
+// api types. If any errors occur during processing, they're returned here.
 func (f *StorageTypeFactory) Types() (astmodel.Types, error) {
-	if len(f.outputTypes) == 0 {
+
+	if !f.processed {
 		err := f.process()
+		f.processed = true
 		if err != nil {
 			return nil, err
 		}
 	}
 
-	return f.outputTypes, nil
+	return f.types, nil
 }
 
+// process carries out our transformations
+// Each step reads from outputTypes and puts the results back in there
```

Comment:
minor: There is no more `outputTypes`

Reply:
there is no more `storage_type_factory` too! 😁